### PR TITLE
Extend transactions to cover catalog name to connector instance

### PR DESCRIPTION
**Pull request by dain**

### CODE REVIEW

File: presto-benchmark/src/main/java/com/facebook/presto/benchmark/BenchmarkSuite.java [hunk 1]
```diff
@@ -47,11 +50,7 @@
                 new HashBuildBenchmark(localQueryRunner),
                 new HashJoinBenchmark(localQueryRunner),
                 new HashBuildAndJoinBenchmark(localQueryRunner.getDefaultSession(), localQueryRunner),
-                new HashBuildAndJoinBenchmark(
```
Comment (electrum) [inline presto-benchmark/src/main/java/com/facebook/presto/benchmark/BenchmarkSuite.java]: This is just extracting a variable?

File: presto-client/src/main/java/com/facebook/presto/client/StatementClient.java [hunk 2]
```diff
@@ -136,7 +136,9 @@ private Request buildQueryRequest(ClientSession session, String query)
             builder.setHeader(PrestoHeaders.PRESTO_SCHEMA, session.getSchema());
         }
         builder.setHeader(PrestoHeaders.PRESTO_TIME_ZONE, session.getTimeZoneId());
-        builder.setHeader(PrestoHeaders.PRESTO_LANGUAGE, session.getLocale().toLanguageTag());
+        if (session.getLocale() != null) {
```
Comment (electrum) [inline presto-client/src/main/java/com/facebook/presto/client/StatementClient.java]: Why is this needed as part of this commit?

File: presto-main/src/main/java/com/facebook/presto/FullConnectorSession.java [hunk 3]
```diff
@@ -109,13 +114,13 @@ public long getStartTime()
     }
 
     @Override
-    public <T> T getProperty(String name, Class<T> type)
+    public <T> T getProperty(String propertyName, Class<T> type)
     {
         if (properties == null) {
-            throw new PrestoException(StandardErrorCode.INVALID_SESSION_PROPERTY, "Unknown session property " + name);
+            throw new PrestoException(StandardErrorCode.INVALID_SESSION_PROPERTY, "Unknown session property " + catalog + "." + propertyName);
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/FullConnectorSession.java]: Static import and use `format`:
```java
throw new PrestoException(INVALID_SESSION_PROPERTY, format("Unknown session property: %s.%s", catalog, propertyName));

File: presto-main/src/main/java/com/facebook/presto/Session.java [hunk 4]
```diff
@@ -92,11 +98,18 @@ public Session(
...
-        ImmutableMap.Builder<String, Map<String, String>> catalogPropertiesBuilder = ImmutableMap.<String, Map<String, String>>builder();
-        catalogProperties.entrySet().stream()
+        ImmutableMap.Builder<ConnectorId, Map<String, String>> catalogPropertiesBuilder = ImmutableMap.builder();
+        connectorProperties.entrySet().stream()
                 .map(entry -> Maps.immutableEntry(entry.getKey(), ImmutableMap.copyOf(entry.getValue())))
                 .forEach(catalogPropertiesBuilder::put);
-        this.catalogProperties = catalogPropertiesBuilder.build();
+        this.connectorProperties = catalogPropertiesBuilder.build();
+
+        ImmutableMap.Builder<String, Map<String, String>> unprocessedCatalogPropertiesBuilder = ImmutableMap.builder();
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/Session.java]: This would be a bit easier to read with a helper method:
```java
this.unprocessedCatalogProperties = deepCopy(unprocessedCatalogProperties);

File: presto-main/src/main/java/com/facebook/presto/Session.java [hunk 5]
```diff
@@ -92,11 +98,18 @@ public Session(
...
+        unprocessedCatalogProperties.entrySet().stream()
+                .map(entry -> Maps.immutableEntry(entry.getKey(), ImmutableMap.copyOf(entry.getValue())))
+                .forEach(unprocessedCatalogPropertiesBuilder::put);
+        this.unprocessedCatalogProperties = unprocessedCatalogPropertiesBuilder.build();
+        checkArgument(!transactionId.isPresent() || unprocessedCatalogProperties.isEmpty(), "Catalog session properties can not be set if there is an open transaction");
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/Session.java]: "cannot"

File: presto-main/src/main/java/com/facebook/presto/Session.java [hunk 6]
```diff
@@ -205,10 +223,43 @@ public String getPreparedStatementFromExecute(Execute execute)
...
+
+        for (Entry<String, String> property : systemProperties.entrySet()) {
+            // verify permissions
+            accessControl.checkCanSetSystemSessionProperty(identity, property.getKey());
+
+            // validate session property value
+            sessionPropertyManager.validateSystemSessionProperty(property.getKey(), property.getValue());
+        }
+
+        // Now that there is a transaction the catalog name can be resolved to a connector, and the catalog properties can be validated
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/Session.java]: "... transaction, the catalog name ..."

File: presto-main/src/main/java/com/facebook/presto/Session.java [hunk 7]
```diff
@@ -353,6 +406,7 @@ public SessionBuilder setQueryId(QueryId queryId)
 
         public SessionBuilder setTransactionId(TransactionId transactionId)
         {
+            checkArgument(catalogSessionProperties.isEmpty(), "Catalog session properties can not be set if there is an open transaction");
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/Session.java]: "cannot"

File: presto-main/src/main/java/com/facebook/presto/metadata/SessionPropertyManager.java [hunk 8]
```diff
@@ -87,95 +85,127 @@ public void addSystemSessionProperties(List<PropertyMetadata<?>> systemSessionPr
...
-            else {
-                values = session.getSystemProperties();
-            }
-            String value = values.getOrDefault(sessionProperty.getPropertyName(), defaultValue);
-
-            properties.add(new SessionPropertyValue(
+        ImmutableList.Builder<SessionPropertyValue> sessionPropertyValues = ImmutableList.builder();
+        Map<String, String> systemProperties = session.getSystemProperties();
+        for (PropertyMetadata<?> property : systemSessionProperties.values()) {
+            String defaultValue = firstNonNull(property.getDefaultValue(), "").toString();
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/metadata/SessionPropertyManager.java]: This can be `nullToEmpty(property.getDefaultValue())`

(the `toString()` is redundant)

File: presto-main/src/main/java/com/facebook/presto/metadata/SessionPropertyManager.java [hunk 9]
```diff
@@ -87,95 +85,127 @@ public void addSystemSessionProperties(List<PropertyMetadata<?>> systemSessionPr
...
+                    property.isHidden()));
+        }
+
+        for (Entry<String, ConnectorId> entry : new TreeMap<>(catalogs).entrySet()) {
+            String catalog = entry.getKey();
+            ConnectorId connectorId = entry.getValue();
+            Map<String, String> connectorProperties = session.getConnectorProperties(connectorId);
+
+            for (PropertyMetadata<?> property : new TreeMap<>(connectorSessionProperties.get(connectorId)).values()) {
+                String defaultValue = firstNonNull(property.getDefaultValue(), "").toString();
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/metadata/SessionPropertyManager.java]: Same as above

File: presto-main/src/main/java/com/facebook/presto/server/ResourceUtil.java [hunk 10]
```diff
@@ -282,7 +274,8 @@ private static TimeZoneKey getTimeZoneKey(String timeZoneId)
 
     private static Optional<TransactionId> getTransactionId(String transactionId)
     {
-        if (transactionId.toUpperCase().equals("NONE")) {
+        transactionId = trimEmptyToNull(transactionId);
+        if (transactionId == null || transactionId.toUpperCase().equals("NONE")) {
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/ResourceUtil.java]: This could be simplified to
```java
if ("none".equalsIgnoreCase(transactionId)) {

File: presto-main/src/test/java/com/facebook/presto/execution/TestResetSessionTask.java [hunk 11]
```diff
@@ -39,21 +48,39 @@
...
+    private final AccessControl accessControl;
+    private final MetadataManager metadata;
 
     public TestResetSessionTask()
     {
+        CatalogManager catalogManager = new CatalogManager();
+        transactionManager = createTestTransactionManager(catalogManager);
+        accessControl = new AllowAllAccessControl();
+
+        metadata = new MetadataManager(
```
Comment (electrum) [inline presto-main/src/test/java/com/facebook/presto/execution/TestResetSessionTask.java]: This is repeated. It needs a helper method somewhere. Maybe in `MetadataManager`?

File: presto-main/src/main/java/com/facebook/presto/metadata/SessionPropertyManager.java (hunk 8 above)
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/metadata/SessionPropertyManager.java]: getDefaultValue returns `T` not `String`
